relay-server/src/utils/dynamic_sampling.rs

```diff
@@ -0,0 +1,441 @@
+//! Functionality for calculating if a trace should be processed or dropped.
+//!
+use std::convert::TryInto;
+
+use rand::{distributions::Uniform, Rng};
+use rand_pcg::Pcg32;
+use serde::{Deserialize, Serialize};
+
+use relay_common::{ProjectId, ProjectKey, Uuid};
+use relay_filter::GlobPatterns;
+
+/// A sampling rule defined by user in Organization options
+#[derive(Debug, Clone, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct SamplingRule {
+    /// The project ids for which the sample rule applies, empty applies to all
+    #[serde(default)]
+    pub project_ids: Vec<ProjectId>,
+    /// A set of Glob patterns for which the rule applies, empty applies to all
+    #[serde(default)]
+    pub releases: GlobPatterns,
+    /// A set of user_segments for which the rule applies, empty applies to all
+    #[serde(default)]
+    pub user_segments: Vec<String>,
+    /// the sampling rate for trace matching this rule
+    pub sample_rate: f64,
+}
+
+/// Represents the dynamic sampling configuration available to a project.
+/// Note: This comes from the organization data
+#[derive(Debug, Clone, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct SamplingConfig {
+    /// The sampling rules for the project
+    pub rules: Vec<SamplingRule>,
+}
+
+/// TraceContext created by the first Sentry SDK in the call chain
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct TraceContext {
+    /// IID created by SDK to represent the current call flow
+    /// TODO RaduW. Should we deserialize directly into an Uuid ?
+    /// Also since trace_id and public_key are compulsory for tracing functionality should
+    /// they be an option or not ?
```

Comment:
To both questions, the current implementation looks fine to me:

- It is reasonable to deserialize directly into a UUID since the trace id would always be a UUID. We can be more forward compatible, but I think it is fine to assume that a UUID has to be provided by the SDK. 
- The public key indeed required, so there is no point in making it an option.